nectar/downloaders/threaded.py

```diff
@@ -273,119 +274,128 @@ def _fetch(self, request, session):
+
+                    if now - last_update_time >= progress_interval:
+                        last_update_time = now
+                        self.fire_download_progress(report)
+
+                    with self._bytes_lock:
+                        if now - self._time_bytes_this_second_was_cleared >= ONE_SECOND:
+                            self._bytes_this_second = 0
+                            self._time_bytes_this_second_was_cleared = now
+                        self._bytes_this_second += bytes_read
+
+                    if max_speed is not None and self._bytes_this_second >= max_speed:
+                        # it's not worth doing fancier mathematics than this, very
+                        # fine-grained sleep times [1] are not honored by the system
+                        # [1] for example, sleeping the remaining fraction of time
+                        # before this second is up
+                        time.sleep(0.5)
+
+                # guarantee 1 report at the end
+                self.fire_download_progress(report)
+
+            except SkipLocation:
+                _logger.debug("Skipping {url} because {netloc} could not be reached.".format(
+                    url=request.url, netloc=netloc)
+                )
+                report.download_skipped()
+
+            except requests.ConnectionError as e:
+                # retry only if there's indication of connection reset
+                if nretry < DEFAULT_GENERIC_TRIES - 1 and not e.args[1]:
+                    continue
+                _logger.error(_('Skipping requests to {netloc} due to repeated connection'
+                                ' failures: {e}').format(netloc=netloc, e=str(e)))
+                self.failed_netlocs.add(netloc)
+                report.download_connection_error()
+
+            except requests.Timeout:
+                """
+                Handle a timeout differently than a connection error. Do not add
+                to failed_netlocs so that a new connection can be attempted.
+                """
+                _logger.warning("Request Timeout - Connection with {url} timed out.".format(
+                    url=request.url)
+                )
+                report.download_connection_error()
+
+            except DownloadCancelled as e:
+                _logger.info(str(e))
+                report.download_canceled()
+
+            except DownloadFailed as e:
+                # retry only if there's indication of connection reset
+                if nretry < DEFAULT_GENERIC_TRIES - 1 and not e.args[1]:
+                    continue
+                _logger.info('Download failed: %s' % str(e))
+                report.error_msg = e.args[2]
+                report.error_report['response_code'] = e.args[1]
+                report.error_report['response_msg'] = e.args[2]
+                report.download_failed()
+
+            except Exception as e:
+                if nretry < DEFAULT_GENERIC_TRIES - 1:
```

Comment:
I finally managed to reproduce the RST.
It is get caught here.
I inserted some info statements
```
Sep 04 11:33:47 pulp2-nightly-pulp3-source-centos7.fluffy.example.com pulp[8284]: nectar.downloaders.threaded:INFO: PRINT EXCEPTION DETAILS: " e.args"
Sep 04 11:33:47 pulp2-nightly-pulp3-source-centos7.fluffy.example.com pulp[8284]: nectar.downloaders.threaded:INFO: (104, 'Connection reset by peer')
Sep 04 11:33:47 pulp2-nightly-pulp3-source-centos7.fluffy.example.com pulp[8284]: nectar.downloaders.threaded:INFO: PRINT EXCEPTION DETAILS: " response.status_code"
Sep 04 11:33:47 pulp2-nightly-pulp3-source-centos7.fluffy.example.com pulp[8284]: nectar.downloaders.threaded:INFO: 200
```
Given this information I suggest following changes:

```

import errno
----snip---
except Exception as e:
    if nretry < DEFAULT_GENERIC_TRIES - 1 and e.args[1]==errno.ECONNRESET:
                    continue
    _logger.exception(e)
    report.error_msg = str(e)
    report.download_failed()
```

Reply:
Oh, nice! Are you able to reliably reproduce it? so I was wrong about that it caught in DownloadFailed exception, sorry. :(

Reply:
> Oh, nice! Are you able to reliably reproduce it? so I was wrong about that it caught in DownloadFailed exception, sorry. :(

That's ok, actually with your hint i was able to properly spot the place where to insert log statements.
Apparently increasing max_downloads to 30 helps with evoking RST. (it just takes some 20 mins..)